### PR TITLE
[run-clang-tidy,clang-tidy-diff] Accept directory as value for -export-fixes

### LINKED ISSUE #69450: [clang-tidy] run_clang_tidy.py should accept a directory to export clang-tidy fixes
Currently, the only option to get the found issues in a machine readable way is `-export-fixes` which combines the temporary yaml files of each compilation unit into one large one. For large projects, this is not feasable since combining (in the way it is implemented) is very slow.

In the documentation of `-export-fixes`, it says:

> Create a yaml file to store suggested fixes in, which can be applied with clang-apply-replacements.

`clang-apply-replacements` however accepts not a file, but a directory.

I propose to add another parameter `-export-directory` where the yaml-files are saved to. If this parameter is not given, a temporary directory is used that is deleted afterwards.

The use case in our project is that we generate a code quality report (https://docs.gitlab.com/ee/ci/testing/code_quality.html) of the found issues. We need to convert the yaml files into a format that Gitlab accepts.

### PR DESCRIPTION
**Pull request by amgebauer**
Adding an additional parameter to run_clang_tidy.py to accept a directory where the clang-tidy fixes are saved to. This directory can then be used to run `clang-apply-replacements`.

Closes #69450

### CODE REVIEW

**Comment by PiotrZSL** [comment]
There is already --export_fixes that can be used in +- same way.
Maybe better would be to change --export_fixes to take filename or directory (ends with / or exist as directory), and in that case it could save to directory or merge to single file.

**Comment by amgebauer** [comment]
@PiotrZSL Thank you for your feedback. I implemented it the way you suggest. Feel free to leave comments.

File: clang-tools-extra/clang-tidy/tool/run-clang-tidy.py [hunk 1]
```diff
@@ -384,14 +386,26 @@ def main():
...
             args.clang_apply_replacements_binary, "clang-apply-replacements", build_path
         )
 
-    if args.fix or (yaml and args.export_fixes):
-        tmpdir = tempfile.mkdtemp()
+    combine_fixes = False
+    export_fixes_dir = None
+    delete_fixes_dir = True
+    if args.export_fixes is not None:
+        assert not args.export_fixes.endswith(os.path.sep) or os.path.isdir(args.export_fixes), "The export fixes directory does not exist."
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/tool/run-clang-tidy.py:398]: actually we could create directory if it does not exist

File: clang-tools-extra/clang-tidy/tool/run-clang-tidy.py [hunk 2]
```diff
@@ -384,14 +386,26 @@ def main():
...
+
+        if not os.path.isdir(args.export_fixes) and yaml:
+            combine_fixes = True
+        
+        if os.path.isdir(args.export_fixes):
+            export_fixes_dir = args.export_fixes
+            delete_fixes_dir = False
+        
+    if export_fixes_dir is None and (args.fix or combine_fixes):
+        export_fixes_dir = tempfile.mkdtemp()
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/tool/run-clang-tidy.py:408]: set delete_fixes_dir  by default to False, and under this if set it to True

**Review by PiotrZSL — COMMENTED** [review]
Release notes will be needed, in clang-tools-extra/docs/ReleaseNotes.rst.
Additionally probably similar change should be done to clang-tidy-diff.py

File: clang-tools-extra/clang-tidy/tool/clang-tidy-diff.py [hunk 3]
```diff
@@ -327,14 +345,14 @@ def main():
     if yaml and args.export_fixes:
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/tool/clang-tidy-diff.py:345]: if combine_fixes:

**Review by PiotrZSL — APPROVED** [review]
LGTM.
Would be good to test if both scripts are working before merge.

**Comment by amgebauer** [comment]
From my side it is ready to be merged. I don't have the access rights to do so.

**Comment by dyung** [comment]
Hi @amgebauer, not sure if you got the notification, but your change seems to be causing a test failure on two build bots, any idea why?
https://lab.llvm.org/buildbot/#/builders/139/builds/51852
https://lab.llvm.org/buildbot/#/builders/247/builds/10286

**Comment by amgebauer** [comment]
I have an idea about the reason. I will push a fix in a few moments. Thank you for the hint

**Pull request merged**